Leader-Board Implementation

lunayach opened the pull request:
### Description
Implemented the LeaderBoard component. Users are now ranked by their score in their respective badge categories. 

Fixes #131

### Type of Change:


- Code
- User Interface


- Bug fix (non-breaking change which fixes an issue)
- This change requires a documentation update (software upgrade on readme file)
- New feature (a non-breaking change which adds functionality pre-approved by mentors)



### How Has This Been Tested?
Wrote new tests both in Mocha and Karma. Both existing and new tests pass.


### Checklist:


- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged 


- [x] My changes generate no new warnings 
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

![final8](https://user-images.githubusercontent.com/21227893/42422294-d12121ba-8300-11e8-9a37-45021ce6b62f.gif)

Raxerz commented on routes/registration.js:
```diff
@@ -100,7 +101,10 @@ router.post('/', function(req, res) {
                         });
-                    });
+                        activityProgress.create({
+                            user_id: task.dataValues.user_id
```
What about default values for odd one out and drag and drop here?

lunayach replied:
@Raxerz  Default values were set while declaring the model itself in ActivityProgress.js

Raxerz replied:
Okay

Raxerz commented on src/app/introduction/activity-1/activity-1.component.ts:
```diff
@@ -69,8 +70,9 @@ export class HighlightActivityComponent implements OnInit {
-                  this._performanceService.openDialog(currStage); }
+          if (!this.completed) { const currStage = 1;
+                  this._performanceService.openDialog(currStage);
+                  this._leaderBoardService.updateLeaderBoard({activity: 'highlightDefinition', level: 'level1'})}
```
Does the level attribute serve any purpose?

lunayach replied:
@Raxerz  Ya, it serves a purpose. If you see **PointsConfig.ts** I have declared points to be given to activities as per levels. Currently, as all activities have one level, so we are giving *level1* flag. For multiple levels, we can have *level2*, *level3* flags respectively.

Raxerz replied:
Is this being stored in the database?

lunayach replied:
Currently, it's like this.
His score in a particular level is only going to be added when he completes that level for the first time. (not for activities which involve dynamic scores like Odd One Out etc.). So, passing the level flag from the game would ensure that his score is updated for that level.

lunayach commented on src/app/header/header.component.html:
```diff
@@ -10,6 +10,7 @@
     </div>
   <div class="col-md-1">
     <img  [@newBadge]="state" [src]="badge?.image" [matTooltip]="badge?.name" (@newBadge.done) = "checkInitialState($event)" onError="this.src = 'assets/img/badges/default.gif'">
+    <i matTooltip="LeaderBoard" class="fa fa-user icons" (click)="loadLeaderBoard()"></i>
```
@Raxerz  Okay! :+1:

lunayach replied:
@shreyans29 didn't get you here. Can you please elaborate.

Raxerz commented on config/passport.js:
```diff
@@ -51,32 +52,48 @@ module.exports = function(passport, models) {
+                        user_id: user.id
+                    }, defaults: {
+                        user_id: user.id,
+                        dragAndDrop: 0,
```
Are the default values required here since it has been added as default in the models file

lunayach replied:
Yup! removed it.

Raxerz commented on src/app/leader-board/leader-board.component.html:
```diff
@@ -0,0 +1,22 @@
+    </tr>
+    </tbody>
+  </table>
+</div>
```
new line

Raxerz commented on src/app/leader-board/leader-board.component.scss:
```diff
@@ -0,0 +1,21 @@
+
+#table {
+  text-align: center;
+}
```
new line

Raxerz commented on src/app/leader-board/leader-board.component.spec.ts:
```diff
@@ -0,0 +1,47 @@
+    expect(component.badgeName).toEqual(mockData.badge.name);
+
+  });
+});
```
new line

Raxerz commented on src/app/leader-board/leader-board.component.ts:
```diff
@@ -0,0 +1,29 @@
+  close() {
+    this.dialogRef.close();
+  }
+}
```
new line

lunayach commented:
![screenshot from 2018-07-10 00-17-17](https://user-images.githubusercontent.com/21227893/42470030-f02bdca4-83d6-11e8-963a-74e9694c8809.png)
![screenshot from 2018-07-10 00-15-01](https://user-images.githubusercontent.com/21227893/42470044-f8c14f66-83d6-11e8-9f82-14cfa4c27bc2.png)
I have **new line** settings in both the files. For some reason, it's still not adding a new line on **GitHub**.

paavininanda commented on test/test.js:
```diff
@@ -258,7 +258,7 @@ describe('Testing APIs', function() {
                     .expect(200)
                     .end(function(err, res) {
-                        res.body.info.should.equal('PCS003 : Successfully updated user progress status');
+                        res.body.info.should.equal('Successfully updated user progress status');
```
Similarly these changes can be taken up in another PR

shreyans29 replied:
yes..

paavininanda commented on src/app/introduction/activity-2/activity-2.component.ts:
```diff
@@ -15,17 +16,26 @@ export class PcpolicyComponent implements OnInit {
     language: any;
     email: String;
     private _status: object = {stage: 1, activity: 2};
-    public completed = true;
```
How was this true till now..

lunayach replied:
@paavininanda this is Peace Corps Policy component. So, the user just sees it and the level is completed.

lunayach commented on src/app/services/leaderBoard.service.ts:
```diff
@@ -0,0 +1,54 @@
+  showLeaderBoard(data): void {
+    const dialogConfig = new MatDialogConfig();
+    dialogConfig.data = data;
+    dialogConfig.width = '700px';
```
@paavininanda actually *width* here is *minimum width*. So for smaller screens, it adapts automatically.

shreyans29 replied:
okay, can we change the variable name then ?

lunayach replied:
Okay.

lunayach commented on src/app/leader-board/PointsConfig.ts:
```diff
@@ -0,0 +1,51 @@
+    level2: 200
+  },
+  strideAgainst: {
+    level1: 150,
```
Okay. I thought we should vary this as per game difficulty. We may want to discuss this. @paavininanda

shreyans29 replied:
I think its tough to do that given different number of levels.

shreyans29 commented on data/codes.json:
```diff
@@ -3,197 +3,220 @@
 
```
yes.. the registration issue has been solved but we can create a new Issue for it.

lunayach replied:
So, shall I move ahead and revert these changes?

shreyans29 replied:
yes..

lunayach replied:
Okay. They were reverted. :+1:

shreyans29 commented on src/app/introduction/activity-1/activity-1.component.ts:
```diff
@@ -69,8 +70,9 @@ export class HighlightActivityComponent implements OnInit {
                 this.activityComplete = true;
-                if (!this.completed) { const currStage = 1;
-                  this._performanceService.openDialog(currStage); }
+          if (!this.completed) { const currStage = 1;
```
const currStage = 1; new line..

shreyans29 commented on src/app/introduction/activity-1/activity-1.component.ts:
```diff
@@ -69,8 +70,9 @@ export class HighlightActivityComponent implements OnInit {
-                  this._performanceService.openDialog(currStage); }
+          if (!this.completed) { const currStage = 1;
+                  this._performanceService.openDialog(currStage);
+                  this._leaderBoardService.updateLeaderBoard({activity: 'highlightDefinition', level: 'level1'})}
```
move the last brace to new line as well

shreyans29 commented on src/app/leader-board/leader-board.component.scss:
```diff
@@ -0,0 +1,21 @@
+  text-transform: capitalize;
+}
+
+th{
```
spacing

shreyans29 commented on src/app/leader-board/leader-board.component.scss:
```diff
@@ -0,0 +1,21 @@
+@import "../styles/main";
+
+#body{
```
spacing

shreyans29 replied:
this needs to be addressed. #body {

lunayach commented:
Addressed the comments.

shreyans29 commented on routes/api.js:
```diff
@@ -158,6 +159,151 @@ router.patch('/user/badge/update', authenticationHelpers.isAuthOrRedirect, (req,
+    } else {
+        return res.status(400).json({error: errorCode.PCE033.message, code: errorCode.PCE033.code});
+    }
+});
```
Are all  these covered by unit tests ?

lunayach replied:
Done.

shreyans29 commented on src/app/malaria-101/activity-1/activity-1-2.component.ts:
```diff
@@ -108,10 +112,12 @@ export class MalariaLifeCycleComponent implements OnInit {
+          }
           this._sharedData.customSuccessAlert(this.alerts.activitySuccessMsg, this.alerts.activitySuccessTitle);
             this._dashboardService.updateProgressStatus(this._status).subscribe(response => {});
         } else if (arrLength === 6) {
```
You need to add  6 and other constants as const.. It will improve code readability by lot.

shreyans29 commented on src/app/malaria-101/activity-3/activity-3.component.ts:
```diff
@@ -23,8 +24,11 @@ export class OddOneOutComponent implements OnInit {
     private _subscription;
     private _data;
     private _status: object = {stage: 2, activity: 3};
+    private readonly pointsPerCorrectAnswer = 10;
```
follow the CAPITALIZED_WITH_UNDERSCORES naming convention for all static and readonly constants/variable .

shreyans29 commented on src/app/malaria-101/activity-3/activity-3.component.ts:
```diff
@@ -23,8 +24,11 @@ export class OddOneOutComponent implements OnInit {
     private _data;
     private _status: object = {stage: 2, activity: 3};
+    private readonly pointsPerCorrectAnswer = 10;
+    private readonly QUESTION_NUMBER_FIVE = 5;
```
this is the correct way, you have to make the change at other places.

lunayach commented:
Addressed comments.

Raxerz commented on config/passport.js:
```diff
@@ -51,32 +52,46 @@ module.exports = function(passport, models) {
-                    fname: fname,
-                    lname: lname
-                }})
+                email: profile.emails[0].value,
```
Why is the indentation changed?

lunayach replied:
@Raxerz  The indentation was not correct. As locally, **eslint** was giving lint warnings with it. So, I changed that to address the warnings.

Raxerz replied:
Oh okay

Raxerz commented on config/passport.js:
```diff
@@ -51,32 +52,46 @@ module.exports = function(passport, models) {
-                            stage: 0,
-                            activity: 0
-                        }})
+                        user_id: user.id,
```
Same problem with the indentation.

lunayach replied:
Same reason as above.

Raxerz commented on database/models/activityProgress.js:
```diff
@@ -0,0 +1,20 @@
+            unique: true
+        }
+    }, {
+        tableName: 'activityProgress'
```
To maintain consistency, rename the table with underscore separation like 'activity_progress'

Raxerz commented on routes/api.js:
```diff
@@ -72,7 +73,7 @@ router.use(function(req, res, next) {
  */
 router.get('/getUserInfo', authenticationHelpers.isAuthOrRedirect, (req, res) => {
-    res.status(200).json({user: req.user});
+    res.status(200).json({user: req.user.email});
```
Why was this change made? This might affect the file like navbar.component.ts which consumes this API.

lunayach replied:
Actually, this API  was not used anywhere in the application until now. In navbar component, we use **get Username** API. 
From this API I get the user's email which I use in Leaderboard component to highlight the current user.

Raxerz commented on routes/api.js:
```diff
@@ -158,6 +159,151 @@ router.patch('/user/badge/update', authenticationHelpers.isAuthOrRedirect, (req,
+    if (req.body) {
+        progress.findAll({
+            include: [localUser],
+            where: {badge: req.body.badge},
```
Why is badge being used in where clause?
Are you ranking users based on a particular badge?

lunayach replied:
Exactly! @Raxerz, users would be ranked badge wise.

Raxerz replied:
@lunayach this is a good idea but was this implementation discussed with the team?
@Buddhiprabha @shreyans29 @paavininanda should this be badge based or for the entire set of activities/games?

lunayach replied:
This was the idea I suggested in the proposal. This would ensure fair scoring between those who have played more stages than those who are still in earlier levels.

Raxerz replied:
Okay

Raxerz replied:
@lunayach what happens when there are no matching badges for any of the users?
How will the leaderboard be displayed? I tried this scenario and the leaderboard dialog was not displayed

lunayach replied:
@Raxerz  This scenario is not possible because all user must be having one badge. That is, even for a new user he is having a beginner badge.
The reason for not showing up of dialogue could be the columns of badge missing. You may try to reset database and try again.

shreyans29 commented on src/app/leader-board/leader-board.component.scss:
```diff
@@ -0,0 +1,31 @@
+    color: $black;
+}
+
+th{
```
spacing here too..

lunayach replied:
Done.

Raxerz commented on src/app/malaria-101/activity-3/activity-3.component.ts:
```diff
@@ -23,8 +24,11 @@ export class OddOneOutComponent implements OnInit {
     private _data;
     private _status: object = {stage: 2, activity: 3};
+    private readonly POINTS_PER_CORRECT_ANSWER = 10;
+    private readonly QUESTION_NUMBER_FIVE = 5;
```
Isn't it better to change the variable name to LAST_QUESTION or something

lunayach replied:
Done.

Raxerz commented on src/app/services/leaderBoard.service.ts:
```diff
@@ -0,0 +1,56 @@
+  private _getLeaderBoard = this._baseAPIUrl + 'leaderBoard';
+  private _getScore = this._baseAPIUrl + 'user/score';
+  private _updateScore = this._baseAPIUrl + 'user/score/update';
+  private readonly dialogMinimumWidth = '700px';
```
can be a constant?

lunayach replied:
@Raxerz are you talking about **dialogMinimumWidth** here?

Raxerz replied:
Yes. Its a constant right?

shreyans29 replied:
yep, its read only has to be. Naming convention..

lunayach replied:
Okay! Done.

Raxerz commented on src/app/introduction/activity-1/activity-1.component.ts:
```diff
@@ -23,8 +24,9 @@ export class HighlightActivityComponent implements OnInit {
     public language: any;
     public alerts: any;
     public completed = false;
+    private readonly currStage = 1;
```
Constant naming convention to be followed, please change it in the other files as well.

shreyans29 commented on src/app/introduction/activity-3/activity-3.component.ts:
```diff
@@ -368,7 +370,8 @@ export class PicturePuzzleComponent implements OnInit {
           this._performanceService.openDialog(currStage);
           this._badgeService.updateBadgeNumber(badgeNumber).subscribe(res => res);
-        }
+          this._leaderBoardService.updateLeaderBoard({activity: 'picturePuzzle', level: 'level1'})
```
even here ?

shreyans29 commented on src/app/malaria-101/activity-2/activity-2.component.ts:
```diff
@@ -16,6 +16,8 @@ export class DragdropComponent implements OnInit {
     public dropCheckDont: String;
     language: any;
+    private score = 0;
+    private readonly pointsPerCorrectAnswer = 10;
```
yes please change it to CAPS_WITH_UNDERSCORE

lunayach replied:
Done.

shreyans29 commented on src/app/meds-n-labels/activity-1/activity-1.component.ts:
```diff
@@ -294,11 +296,13 @@ export class MatchmedsComponent implements OnInit {
+                          }
+                              this._dashboardService.updateProgressStatus(this._status).subscribe(response => {});
+                              this._infokitService.activateinfokit('match_meds').subscribe(res => {});
+                              this._leaderBoardService.updateLeaderBoard({activity: 'matchMeds', level: 'level1'});
```
same thing here..  no hardcoded strings.

shreyans29 commented on src/app/meds-n-labels/activity-2/activity-2.component.ts:
```diff
@@ -168,9 +170,9 @@ export class MemoryGameComponent implements OnInit {
                 this.activityComplete = true;
                 if (!this.completed) {
                   const badgeNumber = 3;
```
CAPS_WITH_UNDERSCORE naming convention

lunayach replied:
Done.

lunayach commented on src/app/leader-board/leader-board.component.html:
```diff
@@ -0,0 +1,23 @@
+    </thead>
+    <tbody *ngFor="let user of data; index as i">
+    <tr>
+      <td [ngClass] = "{ 'user' : user.userEmail === userEmail } ">{{i + 1}}</td>
```
Here we are using the **user** from the **data** array in the For loop. It becomes convenient to check the condition of active user, while we are binding it in the template. 
It can be moved to the .ts file, then I check for this condition and associate some value to a some new variable, and then I again check equality of the variable, while binding it. 
So some condition needs to checked in the template in either case, as we want to control styling of the particular division.

Raxerz replied:
Can't we have a function that takes it as a parameter in the ts file and returns the appropriate value?

shreyans29 replied:
I think what @Raxerz is suggesting will work and is a good way.

lunayach replied:
Okay! Got it. Done.

lunayach commented on src/app/malaria-101/activity-3/activity-3.component.ts:
```diff
@@ -23,8 +24,11 @@ export class OddOneOutComponent implements OnInit {
 
-    public activityComplete;
+
+  public activityComplete;
```
Done.

lunayach commented on src/app/meds-n-labels/activity-2/activity-2.component.ts:
```diff
@@ -26,6 +27,7 @@ export class MemoryGameComponent implements OnInit {
     private _activity: number;
     private _stage: number;
     private _baseImgPath = '../../assets/img/Memory-game/';
+    private readonly currStage = 7;
```
done.

lunayach commented on src/app/meds-n-labels/activity-2/activity-2.component.ts:
```diff
@@ -26,6 +27,7 @@ export class MemoryGameComponent implements OnInit {
     private _activity: number;
     private _stage: number;
     private _baseImgPath = '../../assets/img/Memory-game/';
+    private readonly currStage = 7;
```
Done.

lunayach commented on src/app/prevention/stop-the-breed/stop-the-breed.component.ts:
```diff
@@ -46,8 +48,11 @@ export class StopTheBreedComponent implements OnInit {
-      if (!this.completed) { const currStage = 8;
-        this._performanceService.openDialog(currStage); }
+      if (!this.completed) {
+        const currStage = 8;
```
Done.

lunayach commented on src/app/introduction/activity-1/activity-1.component.ts:
```diff
@@ -69,8 +71,10 @@ export class HighlightActivityComponent implements OnInit {
-                  this._performanceService.openDialog(currStage); }
+              if (!this.completed) {
+                this._performanceService.openDialog(this.currStage);
+                this._leaderBoardService.updateLeaderBoard({activity: 'highlightDefinition', level: 'level1'})
```
@shreyans29 I changed the hardcoded string of activities in all the files.
But, since **level** is dynamic and would depend on different levels, so this would become a global variable when we would have **muliple levels** PR merged.

lunayach commented:
Addressed comments and LeaderBoard looks like this:
![screenshot from 2018-07-22 02-33-04](https://user-images.githubusercontent.com/21227893/43037553-e8d18ea0-8d2b-11e8-977c-d72559b7ea3e.png)

shreyans29 commented on src/app/malaria-101/activity-2/activity-2.component.html:
```diff
@@ -8,7 +8,7 @@ <h2 class="sub-heading-1 text-center"> {{language?.heading}} </h2>
       <div class="row">
         <div class="col-xs-4 col-xs-offset-2">
-          <div dnd-droppable [className]="'panel panel-success ' + dropCheckDo" (onDropSuccess)="addTobox($event,box1)" (onDragOver)="checkDrop($event,box1)" (onDragLeave)="reset()">
+          <div dnd-droppable [className]="'panel panel-success '" (onDropSuccess)="addTobox($event,box1)"  (onDragLeave)="reset()">
```
got it..

lunayach commented:
Addressed comments.
